warc2zim/main.py

```diff
@@ -0,0 +1,309 @@
+            return self.record.rec_headers['Content-Type']
+
+    def get_url(self):
+        return 'A/' + self.url.split('//', 2)[1]
+
+    def get_mime_type(self):
+        return self.mime
+
+    def get_data(self):
+        return Blob(self.payload)
+
+
+# ============================================================================
+class RWPStaticArticle(BaseArticle):
+    def __init__(self, prefix, filename):
+        super(RWPStaticArticle, self).__init__()
+        self.prefix = prefix
+        self.filename = filename
+
+        try:
+            resp = requests.get(self.prefix + filename)
+            self.content = resp.content
+            self.mime = resp.headers.get('Content-Type').split(';')[0]
+        except Exception as e:
+            logger.error(e)
+            logger.error('Unable to load replay system file: {0}'.format(self.prefix + filename))
+            raise
+
+    def get_url(self):
+        return 'A/' + self.filename
+
+    def get_mime_type(self):
+        return self.mime
+
+    def get_data(self):
+        return Blob(self.content)
+
+
+# ============================================================================
+class RWPViewerArticle(BaseArticle):
+    def __init__(self, filename, main_url):
+        super(RWPViewerArticle, self).__init__()
+        self.filename = filename
+        self.main_url = main_url
+
+    def get_url(self):
+        return 'A/' + self.filename
+
+    def get_mime_type(self):
+        return 'text/html'
+
+    def get_data(self):
+        content = """
+<html>
+  <head>
+    <style>
+    body {{
+      width: 100%;
+      height: 100%;
+      overflow-y: hidden;
+      margin: 0px;
+      padding: 0px;
+    }}
+    </style>
+    <script src="./ui.js"></script>
+  </head>
+  <body>
+    <replay-web-page
+     source="proxy:../"
+     config='{{"type": "kiwix"}}'
+     replayBase="./"
+     url="{0}"
+     embed="replayonly"
+     deepLink="true"
+     />
+  </body>
+</html>
+""".format(self.main_url)
+
+        return Blob(content.encode('utf-8'))
+
+
+# ============================================================================
+class WARC2Zim:
+    REPLAY_STATIC_FILES = ['index.html', 'ui.js', 'sw.js']
+
+    def __init__(self, args):
+        logging.basicConfig(format='[%(levelname)s] %(message)s')
+        if args.verbose:
+            logger.setLevel(logging.DEBUG)
+        else:
+            logger.setLevel(logging.INFO)
+
+        self.indexed_urls = set({})
+        self.name = args.name
+        if not self.name:
+            self.name, ext = os.path.splitext(args.inputs[0])
+            self.name += '.zim'
+
+        self.inputs = args.inputs
+        self.replay_viewer_source = args.replay_viewer_source
+        self.main_url = args.main_url
+
+        self.replay_articles = []
+        self.revisits = {}
+
+    def run(self):
+        for filename in self.REPLAY_STATIC_FILES:
+            self.replay_articles.append(RWPStaticArticle(self.replay_viewer_source, filename))
```

Comment:
This this was because the RWPStaticArticle can raise an exception in case file isn't readable, and in that case should exit and not write anything.. If it is inside Creator, looks like creator will still write an empty zim file, and error code is 0..
But could change if you prefer..

Reply:
Updated the test to ensure no zim is created in case of an exception there. The change would then break that test.

Reply:
Yeah it's a limitation of libzim/pylibzim. See https://github.com/openzim/libzim/issues/347 and the [workaround in zimscraperlib](https://github.com/openzim/python_scraperlib/blob/500e44509f0a94911010290c90701db8a72a504f/src/zimscraperlib/zim/creator.py#L147)

Reply:
But it's hardly testable so I'll leave it to you.